### ISSUE TITLE
Assertion Failure on mouse hovering in TBaseVirtualTree.DoGetImageText

### ISSUE THREAD

**Issue by the-Arioch** [JAM-Software/Virtual-TreeView#710]
A stub to get a reference number.
Would fill later with more details

Core issue is mismatch of expectations between `TBTV.GetImageSize` (and its first statement obviously felt victim to copy-paste) and `TBTV.DoGetImageIndex` it calls, and `OnGetImageIndex` and `OnGetImageIndex` events called by the latter.

Basically the situation "there is no image to be had" should be expected and supported, rathe r than asserted. 
We can never expect the users of the library (application developers) being 100% aware and accurate, especially when compiling their legacy code over yet-developing library.

Basically I feel the whole idea bout images should be described in human readable text (if it was not already, in VTV5 PDF or somewhere else), so when seeing mismatch inside VTV code we knew which part is faulty and which part is reference.

**Comment by the-Arioch** [comment]
`TVTImageKind` has four distinct values. However as `TBTV.DoGetImageIndex` body shows, we actually have two states there: `ikState` or anything else.

I will assume here forth that is the correct idea, if not the rest would not apply.

So, the `TBTV.GetImageSize` first line should be checking exactly this - whether it is `ikState` or not.

Additionally, after calling `OnGet....` we should never call assertions.
We should dictate explicitly to developers which value they are allowed or forbidden to feed us.
Unless it contradicts the intentions, we I believe should be permissive and ignore incorrect values.

**Comment by the-Arioch** [comment]
We have the following three events now.

```
  TVTGetImageEvent = procedure(Sender: TBaseVirtualTree; Node: PVirtualNode; Kind: TVTImageKind; Column: TColumnIndex;
    var Ghosted: Boolean; var ImageIndex: TImageIndex) of object;
  TVTGetImageExEvent = procedure(Sender: TBaseVirtualTree; Node: PVirtualNode; Kind: TVTImageKind; Column: TColumnIndex;
    var Ghosted: Boolean; var ImageIndex: TImageIndex; var ImageList: TCustomImageList) of object;
  TVTGetImageTextEvent = procedure(Sender: TBaseVirtualTree; Node: PVirtualNode; Kind: TVTImageKind; Column: TColumnIndex;
    var ImageText: string) of object;
```

We have to ask ourselves (or app developers would)

There is some "default" imagelist that is assumed by VTV lib at this specific moment in time.
(it may be nil also)

1) what if ImageIndex is set, but is outside of the default imagelist content?
1a) should be ignored and treated like "no image" 
1b) should be given an explicit error "wrong image selected"

2) what if ImageIndex is set, yet default ImageList is nil?
2a) should be ignored and treated like "no image" 
2b) should be given an explicit error "no sources to retrieve images from"

3) what if `OnGetImageEx` event returns nil imagelist (while ImageIndex gets set ) ?
We here are getting back to the prior questions and options. But we should define our angle of view first what we work with if event-provided imagelist is nil ?
3a) fall down to default VTV-chosen imagelist ?
3b) assume default Imagelist is nil for the context, no fallback ?

This can also be rephrased as
3a) user is given a possibility to override the imagelist, but is not required to
3b) user is asked to give or to not to give the imagelist, and VTV would not try to offer tis own imagelists.

Current implementation of `DoGetImageIndex` sports 3b approach via pre-setting the ImageList argument parameter. The developer can just ignore it and leave it be, or can explicitly override it with nil or any other value.

4) how does ImageText intertwine with those choices, can it be - like it is in HTML IMG ALT - a stub, a substituting placeholder for an image? Say, if we got a set ImageIndex value but due to 1) or 2) conditions can not fetch the actual image - can we consider this text a substitute for the image (thus it is okay to fail getting image as long as we get the text and can turn that text into the image)? or should we throw the error even if the text was filled (image is image, text is text, one does not substitute for another)?

**Comment by the-Arioch** [comment]
Like I said above, personally I bias towards permissive libs, so as not as it has value to be picky and throw errors (which may be opted in by developer by some TreeOptions flag for example), the tree should consider "erroneous image" cases as "no image" ones. It would possibly make debugging of applications harder, but would hopefully provide for smoother upgrade path. 

But I do not insist here. Just to decide upon something I set for myself for a while that

1a, 2a, 3b, and 4 as (no relations, no substitutions), just because I do not know what that ImageText is and what it was supposed to be.

**Comment by the-Arioch** [comment]
Now if I only new how to make pull-request, that would be isolated and would not implicitly include ALL my future commits :-(

Maybe i need some *rebase* voodoo or something.

**Comment by the-Arioch** [comment]
Perhaps it is your time to fetch and merge from me, cause making proper pull-request on GitHub seems not possible to me

**Comment by joachimmarder** [comment]
> So, the TBTV.GetImageSize first line should be checking exactly this - whether it is ikState or not.

It seems that there is a copy&paste failure in this if-statement. We should correct that.

> Additionally, after calling OnGet.... we should never call assertions.

This was a result of #629 and lots of similar support questions I received before. After I added those assertions, these kind of support questions was drastically reduced.

**Comment by the-Arioch** [comment]
1)  i don't see how this fix fixes it, because ASSERTS are expected to be ignored in typical production code, thus it merely propagates incorrectly set values further down the VTV internals, and nothing more. 
Now that i almost made owner of our legacy project i prefer assertions always on for a while. If they were off, then the assertions would besilently ignored, just like those users complain.

2) it also breaks the production code, that for years worked with VTV before your XE3 restart. So you clearly steered apart if of prior concepts and assumptions. IOW to sooth newbies your break compatibility, which would mean current users would have less incentive to update VTV they already use.

Let's think about something like `VTV.TreeOptions.DebugOptions` which would tune VTV's attitude between being more permissive and being more paranoid. Like i suggested in the comments about React... procedure.

Just the same as with that DFM fix, i think the should be no breaking changes **by default** , while it might be useful to have **bail-in** settings for those who might prefer them

**Comment by joachimmarder** [comment]
> what if ImageIndex is set, yet default ImageList is nil?
> 2a) should be ignored and treated like "no image"
> 2b) should be given an explicit error "no sources to retrieve images from"

This is clearly an invalid and non-desired state and should be treated as an error, which should be covered by an assertion or exception. 


> what if ImageIndex is set, but is outside of the default imagelist content?
> 1a) should be ignored and treated like "no image"
> 1b) should be given an explicit error "wrong image selected"

Like before, I think 1b) with an assertion would be best, but this not yet exists and needs to be added.


> 3a) user is given a possibility to override the imagelist, but is not required to

This is the status quo and IMHO a good state.

**Comment by joachimmarder** [comment]
> how does ImageText intertwine with those choices,

It is used only in accessibility unit, consumed by screenreaders. The GetImageIndex is so far implemented independently from the actual image. Although I think it makes no sense to have a replacement text for an image if you have no image, I would not spend time on changing this and make it more strict.

**Comment by joachimmarder** [comment]
> i don't see how this fix fixes it

It worked: The number of support questions of people which did not respect the `Kind` parameter in their OnGetImageIndex event handler was drastically reduced and so the change had the desired result.

> because ASSERTS are expected to be ignored in typical production code

I found that raising an exception, so that the error is also in release builds too, was a bit invasive but I am open for putting this to discussion. An advantage of asserts is that they do not cost runtime in release builds but bring up most errors during development.

> So you clearly steered apart if of prior concepts and assumption

Well, sometime one must move on and do it better.

> IOW to sooth newbies your break compatibility, which would mean current users would have less incentive to update VTV they already use.

You need to see the positive side: Previously you obviously calculated  image indexes that have not been used because there was no imagelist. This assertion enabled you to optimize your code after the update.

> Let's think about something like VTV.TreeOptions.DebugOptions

I don't think it makes any sense to duplicate the concept of assertions with such an option.

**Comment by joachimmarder** [comment]
> i think the should be no breaking changes by default

I agree to that, but adding assertions that detect invalid states is not really a breaking change in my opinion. And failed assertions with good messages are easy to understand and to fix. Just because code producing invalid states used to work you cannot guarantee that it will do that for all times.

**Comment by the-Arioch** [comment]
Oh, that was quite a number of points, and obviously i take a different look at them :-)

Well, now i am focused in rolling out our app release, so i think i would continue this discussion later...

Because, well, there are quite a few interconnected points, so i can not put one or two short remarks and go away.

**Comment by joachimmarder** [comment]
No further discussion took place. My conclusion: It makes sense to keep the assertion. Just the fact that old version of Virtual Treeview accepted inconsistent parameter values does not justify to do this forever in future.
